docs/decisions/07-cursor-positions.md

```diff
@@ -0,0 +1,62 @@
+---
+status: draft
+date: 2024-06-21
+---
+# How do we transmit cursor positions?
+
+## Context and Problem Statement
+
+An important feature of collaborative editing is being able to see each other's cursors. How should we implement transmitting them?
+
+## Decision Drivers
+
+Our solution should ideally be:
+
+* Fast
+* Correct (cursors never appear at wrong/shifted locations)
+* Simple (little protocol overhead, little complexity)
+* Take up little disk space
+
+## Considered Options
+
+* {title of option 1}
+* {title of option 2}
+* {title of option 3}
+* … <!-- numbers of options can vary -->
+
+## Decision Outcome
+
+Chosen option: "{title of option 1}", because
+{justification. e.g., only option, which meets k.o. criterion decision driver | which resolves force {force} | … | comes out best (see below)}.
+
+## Pros and Cons of the Options
+
+### Store user state in a map in the main Automerge document
+
+The Automerge document for an Ethersync project could be split into a `files` key (which contains the file contents like before), and a `peers` key (which stores information about the peers' state, like cursor positions).
+
+* Good, because transmiting this information would be trivial, with no work required from us at all. The Automege sync protocol handles it for us.
+* Bad, because this "ephemeral" data would become part of the CRDT, that each peer needs to persist. The persisted data would be larger as a result, and contain data that's completely irrelevant after a short amount of time.
+* Neutral, because it's unclear how to mark user state as "stale". Each peer could delete states after they haven't been updated after (for example) 30 seconds, but could this lead to "fights" between peers that have a connection to the peer, and ones that don't. And if we put the "last seen timestamp" in the CRDT, we assume that all daemons have the same wall time, which isn't necessarily true…
+
+Ideally, user states that have been overwritten anyway could be truncated in the Automerge history, but I don't think that's possible yet.
+
+Because Automerge uses "last writer wins" conflict resolution, the latest peer state should eventually end up in all peers.
+
+### Gossip the user states around, and store them in last-write-wins registers
+
+This would be the appraoch that automerge-repo and Y.js use. Peers send their state information to their connected peers, along with a session ID/user ID and an increasing sequence number, allowing the peers to discard messages they've already seen (and thus, avoiding endless loops). All peers would forward received state messages to all of their other peers, to make sure that the information is available in the whole network.
+
+* Good, because no data accumulates in some CRDT that needs to be persisted forever.
```

Comment:
So the CRDT is more basic in a sense? I'm slightly confused about this statement in combination with the suggestion to use a CRDT crate. We need _some_ conflict resolution, right?

To confirm what you're saying here: Last Write Wins is technically already a CRDT, but not one that _persists its state forever_?

Reply:
That's correct! :)